src/Compilers/Core/Portable/Operations/OperationMapBuilder.cs

```diff
@@ -0,0 +1,62 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class OperationMapBuilder : OperationWalker<Dictionary<SyntaxNode, IOperation>>
+    {
+        internal static readonly OperationMapBuilder Instance = new OperationMapBuilder();
+
+        private OperationMapBuilder() { }
+
+        public override object? Visit(IOperation? operation, Dictionary<SyntaxNode, IOperation> argument)
+        {
+            if (operation == null)
+            {
+                return null;
+            }
+
+            RecordOperation(operation, argument);
```

Comment:
>RecordOperation(operation, argument); [](start = 12, length = 37)

I am not sure why we are making an assumption that every node goes through here. Nothing prevents me from calling specific visit method directly and bypassing this code.

Reply:
This can be somewhat mitigated by making the following changes:
- Do not implement walker on this type directly. Instead, have static method ```public void AddToMap(IOperation? operation, Dictionary<SyntaxNode, IOperation> argument)```
- The walker type should be private nested type with a singleton that is never exposed outside.
- AddToMap calls Visit on the singleton.

This way we are going to have a full control over how the visitor is used, but we will continue to be vulnerable to what base implementation is doing.

---
In reply to: [526924753](https://github.com/dotnet/roslyn/pull/49453#discussion_r526924753) [](ancestors = 526924753)

Reply:
I'll make those suggestions. Note that `OperationWalker` is a public type, so we need to keep the behavior consistent to avoid breaking our consumers as well.

---
In reply to: [526937934](https://github.com/dotnet/roslyn/pull/49453#discussion_r526937934) [](ancestors = 526937934,526924753)